sympy/solvers/solveset.py

```diff
@@ -3025,6 +3025,23 @@ def _extract_main_soln(sym, sol, soln_imageset):
             # if there is union, then need to check
             # Complement, Intersection, Imageset.
             # Order should not be changed.
+            if isinstance(sol, ConditionSet):
+                sol = sol.args[2]
```

Comment:
```suggestion
                sol = sol.base_set
```
But let's say you have `sol = ConditionSet(x, x>3, Union(Interval(0,5), Interval(10,20)))` then the `base_set` is
`Union(Interval(0, 5), Interval(10, 20))`...is the next `if isinstance(sol, Union)` going to do the right thing with that? We would just be re-adding the intervals to `sol`.

One way of documenting what is happening is to show example solutions being "extracted" from a few different sorts of inputs. This function really doesn't tell what it is suppose to be doing, what does it mean to "Separate the Complements, Intersections, ImageSet lambda expr and its base_set."?
`

Reply:
> But let's say you have sol = ConditionSet(x, x>3, Union(Interval(0,5), Interval(10,20)))

I think this will never happen since it's only concerned about the `Reals` and `Complexes` and `nonlinsolve` don't accept the `domain` as an input. So I think there is no possibility of such intervals.
(see line `L-3337` below)
https://github.com/sympy/sympy/blob/f4c2dce477141ac1b79fea79ed71692988bbb4bf/sympy/solvers/solveset.py#L3327-L3333

Reply:
> One way of documenting what is happening is to show example solutions being "extracted" from a few different sorts of inputs. 

Yes, it should have examples in order to have a better understanding.


> This function really doesn't tell what it is suppose to be doing, what does it mean to "Separate the Complements, Intersections, ImageSet lambda expr and its base_set."?

I think it's extracting solutions from different classes of sets.